File: app/src/main/java/com/sjn/taggingplayer/ui/fragment/PagerFragment.java
package com.sjn.taggingplayer.ui.fragment;

import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.sjn.taggingplayer.R;

import java.util.ArrayList;
import java.util.List;

import static com.sjn.taggingplayer.utils.ViewHelper.setFragmentTitle;

public abstract class PagerFragment extends Fragment {
    static final String PAGER_KIND_KEY = "PAGER_KIND";

    private AppBarLayout mAppBarLayout;
    protected ViewPagerAdapter mAdapter;
    protected ViewPager mViewPager;

    abstract protected void setupViewPager(ViewPager viewPager);

    public Fragment getCurrent() {
        if (mAdapter == null) {
            return null;
        }
        return mAdapter.getItem(mViewPager.getCurrentItem());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_pager, container, false);
        mViewPager = (ViewPager) rootView.findViewById(R.id.viewpager);
        setupViewPager(mViewPager);
        TabLayout tabLayout = (TabLayout) rootView.findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(mViewPager);
        mAppBarLayout = (AppBarLayout) rootView.findViewById(R.id.appbar);
        return rootView;
    }

    protected void setTitle(String title) {
        setFragmentTitle(getActivity(), title);
    }

    class ViewPagerAdapter extends FragmentPagerAdapter {
        private final List<Fragment> mFragmentList = new ArrayList<>();
        private final List<String> mFragmentTitleList = new ArrayList<>();

        ViewPagerAdapter(FragmentManager manager) {
            super(manager);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragmentList.get(position);
        }

        @Override
        public int getCount() {
            return mFragmentList.size();
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitleList.get(position);
        }

        void addFragment(Fragment fragment, String title) {
            mFragmentList.add(fragment);
            mFragmentTitleList.add(title);
        }
    }


}

File: app/src/main/java/com/sjn/stamp/ui/fragment/media/PagerFragment.kt
package com.sjn.stamp.ui.fragment.media

import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.ViewPager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.sjn.stamp.R
import com.sjn.stamp.utils.ViewHelper
import java.util.*

abstract class PagerFragment : FabFragment() {

    private var viewPager: ViewPager? = null
    protected var viewPagerAdapter: ViewPagerAdapter? = null
    protected var fragments: List<PageFragmentContainer>? = null

    val current: Fragment?
        get() = viewPager?.let {
            viewPagerAdapter?.getItem(it.currentItem)
        }

    internal abstract fun setUpFragmentContainer(): List<PageFragmentContainer>

    private fun setupViewPager(viewPager: ViewPager, savedInstanceState: Bundle?) {
        fragments ?: run {
            fragments = setUpFragmentContainer()
        }
        viewPagerAdapter = ViewPagerAdapter(childFragmentManager)
        fragments?.let {
            for (fragmentContainer in it) {
                fragmentContainer.apply {
                    findOrCreate(savedInstanceState, childFragmentManager)
                    fragmentContainer.fragment?.let {
                        viewPagerAdapter?.addFragment(it, fragmentContainer.label)
                    }
                }
            }
        }
        viewPager.adapter = viewPagerAdapter
    }

    class PageFragmentContainer(var label: String, private var fragmentHint: String, private var creator: Creator) {
        var fragment: Fragment? = null
        var tag: String = label + fragmentHint

        fun findOrCreate(savedInstanceState: Bundle?, fragmentManager: FragmentManager) {
            if (fragment != null) {
                return
            }
            fragment = if (fragmentManager.findFragmentByTag(tag) != null) {
                fragmentManager.getFragment(savedInstanceState, tag)
            } else {
                creator.create(fragmentHint)
            }
        }

        interface Creator {
            fun create(fragmentHint: String): Fragment
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_pager, container, false)
        viewPager = rootView.findViewById<View>(R.id.viewpager) as ViewPager
        viewPager?.let {
            setupViewPager(it, savedInstanceState)
        }
        val tabLayout = rootView.findViewById<View>(R.id.tabs) as TabLayout
        tabLayout.setupWithViewPager(viewPager)
        return rootView
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        fragments?.let {
            it.filter { it.fragment?.isAdded == true }
                    .forEach { childFragmentManager.putFragment(outState, it.tag, it.fragment) }

        }
    }

    protected fun setTitle(title: String) {
        ViewHelper.setFragmentTitle(activity, title)
    }

    inner class ViewPagerAdapter(manager: FragmentManager) : FragmentPagerAdapter(manager) {
        private val fragmentList = ArrayList<Fragment>()
        private val fragmentTitleList = ArrayList<String>()

        override fun getItem(position: Int): Fragment {
            return fragmentList[position]
        }

        override fun getCount(): Int {
            return fragmentList.size
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return fragmentTitleList[position]
        }

        fun addFragment(fragment: Fragment, title: String) {
            fragmentList.add(fragment)
            fragmentTitleList.add(title)
        }
    }

    companion object {
        internal const val PAGER_KIND_KEY = "PAGER_KIND"
    }


}